source/adapters/opencl/adapter.cpp

```diff
@@ -8,16 +8,49 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "adapter.hpp"
 #include "common.hpp"
-#include "logger/ur_logger.hpp"
+#include "ur/ur.hpp"
+
+#ifdef _MSC_VER
+#include <Windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+ur_adapter_handle_t_::ur_adapter_handle_t_() {
+#ifdef _MSC_VER
+  // Loading OpenCL.dll increments the libraries internal reference count.
+  auto handle = LoadLibraryA("OpenCL.dll");
```

Comment:
I think we could that this handle is not `NULL` and if so, we print an error that it failed to find the opencl module.

Reply:
Is this guaranteed to bring the same `opencl.dll` that was used in the adapter initially. I am just concerned that this will load a system `opencl.dll` and the adapter is already using a local fetch of the opencl.dll that lives in `build/bin` folder and they would have a different version (not still sure if that would cause also a harm if that is the case)?

Reply:
> I think we could that this handle is not NULL and if so, we print an error that it failed to find the opencl module.

We've already loaded OpenCL.dll by this point, the adapter would fail to load if it were not present and this code would not be capable of being executed. We're just asking for the handle to the already loaded OpenCL.dll.

> Is this guaranteed to bring the same opencl.dll that was used in the adapter initially.

From the remarks section in [LoadLibraryA](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya):

> If the specified module is a DLL that is not already loaded for the calling process, the system calls the DLL's [DllMain](https://learn.microsoft.com/en-us/windows/desktop/Dlls/dllmain) function with the DLL_PROCESS_ATTACH value.

Thie implication being that if its already loaded it does not load the library a different library of the same name.

Reply:
Cool, that make sense. Thanks for clarification!